src/Sentry/MetricAggregator.cs

```diff
@@ -0,0 +1,466 @@
+
+    /// <inheritdoc cref="IMetricAggregator.FlushAsync"/>
+    public async Task FlushAsync(bool force = true, CancellationToken cancellationToken = default)
+    {
+        try
+        {
+            await _flushLock.WaitAsync(cancellationToken).ConfigureAwait(false);
+
+            foreach (var key in GetFlushableBuckets(force))
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                _options.LogDebug("Flushing metrics for bucket {0}", key);
+                if (!Buckets.TryRemove(key, out var bucket))
+                {
+                    continue;
+                }
+
+                _captureMetrics(bucket.Values);
+                _options.LogDebug("Metric flushed for bucket {0}", key);
+            }
+
+            foreach (var (timestamp, locations) in FlushableLocations())
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                _options.LogDebug("Flushing code locations: ", timestamp);
+                var codeLocations = new CodeLocations(timestamp, locations);
+                _captureCodeLocations(codeLocations);
+                _options.LogDebug("Code locations flushed: ", timestamp);
+            }
+
+            ClearStaleLocations();
+        }
+        catch (OperationCanceledException)
+        {
+            _options.LogInfo("Shutdown token triggered. Exiting metric aggregator.");
+        }
+        catch (Exception exception)
+        {
+            _options.LogError(exception, "Error processing metrics.");
+        }
+        finally
+        {
+            _flushLock.Release();
+        }
+    }
+
+    /// <summary>
+    /// Returns the keys for any buckets that are ready to be flushed (i.e. are for periods before the cutoff)
+    /// </summary>
+    /// <param name="force">Forces all buckets to be flushed, ignoring the cutoff</param>
+    /// <returns>
+    /// An enumerable containing the keys for any buckets that are ready to be flushed
+    /// </returns>
+    internal IEnumerable<long> GetFlushableBuckets(bool force = false)
+    {
+        if (!_buckets.IsValueCreated)
+        {
+            yield break;
+        }
+
+        if (force)
+        {
+            // Return all the buckets in this case
+            foreach (var key in Buckets.Keys)
+            {
+                yield return key;
+            }
+        }
+        else
+        {
+            var cutoff = MetricHelper.GetCutoff();
+            foreach (var key in Buckets.Keys)
+            {
+                var bucketTime = DateTimeOffset.FromUnixTimeSeconds(key);
+                if (bucketTime < cutoff)
+                {
+                    yield return key;
+                }
+            }
+        }
+    }
+
+    private Dictionary<long, Dictionary<MetricResourceIdentifier, SentryStackFrame>> FlushableLocations()
+    {
+        _codeLocationLock.Wait();
+        try
+        {
+            var result = _pendingLocations;
```

Comment:
You can do this atomically with Interlocked afaik, CompareExchange iirc

Reply:
Hm, the `_codeLocationLock` is used elsewhere in the MetricAggregator to lock on different operations against _pendingLocations though. For example, there's this:
https://github.com/getsentry/sentry-dotnet/blob/1c3b2757cf352540472a97794ab7804b062177e2/src/Sentry/MetricAggregator.cs#L210-L214